README.md
<p align="center" >
  <a href="https://100ms.live/">
  <img src="https://raw.githubusercontent.com/100mslive/100ms-ios/featuredSampleApp/100ms-logo.png" title="100ms logo" float=left>
</p>

[![Version](https://img.shields.io/cocoapods/v/HMSVideo.svg?style=flat)](http://cocoapods.org/pods/HMSVideo)
[![Documentation](https://img.shields.io/badge/Read-Documentation-yellow)](https://docs.100ms.live/)
[![Platform](https://img.shields.io/cocoapods/p/HMSVideo.svg?style=flat)](http://cocoapods.org/pods/HMSVideo)
[![License](https://img.shields.io/cocoapods/l/HMSVideo.svg?style=flat)](http://cocoapods.org/pods/HMSVideo)
[![codebeat badge](https://codebeat.co/badges/0ece4c24-7755-4e1d-ae2b-2e33f808ff1e)](https://codebeat.co/projects/github-com-ygit-100ms-ios-main)
[![Slack](https://img.shields.io/badge/Community-Join%20on%20Slack-yellow)](https://join.slack.com/t/100mslive/shared_invite/zt-llwdnz11-vkb2RzptwacwXHO7UeY0CQ)
[![Email](https://img.shields.io/badge/Contact-Know%20More-yellow)](mailto:founders@100ms.live)

## **Setup**

This repository contains example code written in Swift.

### **CocoaPods**

1. Install [CocoaPods 1.7.5 or newer](https://guides.cocoapods.org/using/getting-started.html).
2. Add `pod 'HMSVideo'` to Podfile
3. Run `pod install` from the root directory of this project. CocoaPods will install `HMSVideo.framework` and then set up an `xcworkspace`.
4. Open `HMSVideoExample.xcworkspace`.

Note: You may need to update the CocoaPods [Master Spec Repo](https://github.com/CocoaPods/Specs) by running `pod repo update master` in order to fetch the latest specs for HMSVideo.

## **Quickstart**

### **Running the sample app**

To get started with the sample application follow these steps:

1. Open the `HMSVideo_Example` target from the `HMSVideoExample.xcworkspace` in Xcode and go to `Signing & Capabilities` tab

![Xcode settings](Example/Images/screen1.png?raw=true)

2. Replace the Bundle identifier value with something unique to your org, and select a Team for code signing.

3. Open `MeetingViewController.swift`

![Controller code](Example/Images/screen2.png?raw=true)

4. Replace the tokenServerURL constant value with the correct url ([see instructions below for token api server generation](#token-generation))

5. Build and run the app

![App screen](Example/Images/app.jpg?raw=true)

### Token Generation

You will need to send a token - generated by a combination of `access_key`, `customer_secret`, `customer_id`, `app_id`, `peer_id` and `room_id` when initiating 100ms client

- Follow instructions here: https://100ms.gitbook.io/100ms/helpers/runkit
- Publish your runkit and copy the endpoint URL (click on it and copy the URL that opens).
- Replace the tokenServerURL constant value in `MeetingViewController.swift` with the copied value


# 100ms  iOS (v0.10.0) SDK Documentation

Here you will find everything you need to build experiences with video using 100ms iOS SDK. Dive into our SDKs, quick starts, add real-time video, voice, and screen sharing to your web and mobile applications.

## Pre requisites

- iOS 10.0+
- Xcode 11+

## Installation

### CocoaPods

[CocoaPods](https://cocoapods.org/) is a dependency manager for Cocoa projects. For usage and installation instructions, visit their website. To integrate Brytecam SDK into your Xcode project using CocoaPods, specify it in your `Podfile`:

pod 'HMSVideo', '~> 0.10.0'

## Quick start

Checkout example app at [https://github.com/100mslive/hmsvideo-ios/tree/master/Example](https://github.com/100mslive/hmsvideo-ios/tree/master/Example)

## Concepts

---

- **Room** - A room represents a real-time audio, data, video and/or screenshare session, the basic building block of the Brytecam Video SDK
- **Stream** - A stream represents real-time audio, video and data media streams that are shared to a room
- **Peer/Participant** - A peer represents all participants connected to a room (other than the local participant)
- **Publish** - A local participant can share its audio, video and data tracks by "publishing" its tracks to the room
- **Subscribe** - A local participant can stream any peer's audio, video and data tracks by "subscribing" to their tracks
- **Broadcast** - A local participant can send any message/data to all peers in the room

## Create and instantiate HMSClient (100ms Client)

---

This will instantiate an `HMSClient` object

```swift
//Create an HMSPeer instance for local peer
let peer = HMSPeer(name: userName, authToken: "INSERT TOKEN HERE")

let config = HMSClientConfig()
//config.endpoint = "Override endpoint URL if needed"

//Create a 100ms video client
client = HMSClient(peer: peer, config: config)
```

Click [here](https://www.notion.so/Token-Generation-42b0f9d078224db4bf934608829a8b53) to see how to generate your token

Use `wss://prod-in.100ms.live/ws` as endpoint URL for production and `wss://staging-in.100ms.live/ws` as endpoint URL for staging

## Setup listeners

---

After joining, immediately add listeners to listen to peers joining, new streams being added to the room

```swift
client.onPeerJoin = { (room, peer) in
    // Update UI if needed
}

client.onPeerLeave = { (room, peer) in
    // Update UI if needed
}

client.onStreamAdd = { (room, peer, streamInfo)  in
    // Subscribe to the stream if needed
}

client.onStreamRemove = { (room, peer, streamInfo)  in
    // Remove remote stream view if needed
}

client.onBroadcast = { (room, peer, message) in
    // update UI if needed
}

client.onConnect = { 
		// Client connected, this is a good place to call join(room)
}

client.onDisconnect = { error in 
		// Connection lost or could not be established. 
		// Good place to retry or show an error to the user.
}
```

## Connect

---

After instantiating `HMSClient`, connect to 100ms' server

```swift
//The client will connect to the WebSocket channel provided through the config
client.connect
```

## Join a room

---

```swift
//Pass the unique id for the room here as a String
let room = HMSRoom(roomId: roomName)

client.join(room) { (success, error) in
    //check for error and publish a local stream
}
```

Generate a unique `roomid` for each session to avoid conflicts

## Create and Get local camera/mic streams

---

```swift
//You can set codec, bitrate, framerate, etc here.
let constraints = HMSMediaStreamConstraints()
constraints.shouldPublishAudio = true
constraints.shouldPublishVideo = true
constraints.codec = .VP8
constraints.bitrate = 256
constraints.frameRate = 25
constraints.resolution = .QVGA

let localStream = client.getLocalStream(constraints)
```

Please use the following settings for video that looks good in postcard-sized videos - codec:`VP8`, bitrate `256`, framerate `25`. We will extend this in the future to add more options including front/back camera

Apple requires your app to provide static messages to display to the
user when the system asks for camera or microphone permission: 

If your app uses device cameras, include the  NSCameraUsageDescription
key in your app’s Info.plist file.

If your app uses device microphones, include the  NSMicrophoneUsageDescription
key in your app’s Info.plist file.

For each key, provide a message that explains to the user why your app
needs to capture media, so that the user can feel confident granting
permission to your app.

Important

If the appropriate key is not present in your app’s  Info.plist
file when your app requests authorization or attempts to use a capture
device, the system terminates your app.

## Get local media for screen share

---

**This will not be covered by v0.10 SDK. Coming soon.**

## Display local stream

---

```swift
//The following code is a sample.

//Get the video capturer and video track
let videoCapturer = stream.videoCapturer
let localVideoTrack = stream.videoTracks?.first

//Begin capturing video from the camera
videoCapturer?.startCapture()

//Create a view for rendering video track and add to the UI hierarchy
if let track = localVideoTrack {
    let videoView = HMSVideoView()
		videoView.setVideoTrack(track)
		view.addSubview(videoView)
}
```

## Publish

---

A local participant can share her audio, video and data tracks by "publishing" its tracks to the room

```swift
client.publish(localStream, room: room, completion: { (stream, error) in
    //Handle error if any, update UI if needed
})
```

## Subscribe

---

This method "subscribes" to a peer's stream. This should ideally be called in the `onStreamAdd` listener

```swift
client.subscribe(streamInfo, room: room, completion: { (stream, error) in
	//Handle error if any, update UI if needed
})
```

## Broadcast

---

This method broadcasts a payload to all participants

```swift
client.broadcast(message, room: room, completion: { (stream, error) in
	//Handle error if any, update UI if needed
})
```

## Unpublish local stream

---

```swift
client.unpublish(stream, room: room, completion: { (stream, error) in
	//Handle error if any, update UI if needed
})
```

## Unsubscribe to a peer's stream

---

```swift
client.unsubscribe(stream, room: room, completion: { (stream, error) in
	//Handle error if any, update UI if needed
})
```

## Disconnect client

---

```swift
//The client will disconnect from the WebSocket channel provided
client.disconnect();
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
 
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.10.1] - 2021-02-19

### Added
- Added audio level monitoring
- Added an option to detect screen share stream (isScreen in HMSStreamInfo)
### Changed
- SDK Error codes standardised
- Stability improvements
- Updated example app to stop accessing camera when video muted


HMSVideoExample/Meeting/HMSInteractor.swift
//
//  HMSInteractor.swift
//  HMSVideo_Example
//
//  Created by Yogesh Singh on 26/02/21.
//  Copyright © 2021 100ms. All rights reserved.
//

import Foundation
import HMSVideo

final class HMSInteractor {

    // MARK: - Instance Properties

    private let user: String

    private var updateView: (VideoCellState) -> Void

    private(set) var localPeer: HMSPeer!
    private(set) var client: HMSClient!
    private(set) var room: HMSRoom! {
        didSet {
            let pasteboard = UIPasteboard.general
            pasteboard.string = room.roomId
        }
    }

    private(set) var peers = [String: HMSPeer]()
    private(set) var remoteStreams = [HMSStream]()
    private(set) var videoTracks = [HMSVideoTrack]()

    var model = [VideoModel]()

    private(set) var localStream: HMSStream?
    private(set) var localAudioTrack: HMSAudioTrack?
    private(set) var localVideoTrack: HMSVideoTrack?
    private(set) var videoCapturer: HMSVideoCapturer?

    private(set) var speakerPeerID: String? {
        didSet {
            let streamer = peers.filter { $0.value.peerId == speakerPeerID }
            if let streamID = streamer.keys.first {
                if let track = videoTracks.filter({ $0.streamId == streamID }).first {
                    if speakerVideoTrack?.trackId != track.trackId {
                        speakerVideoTrack = track
                    }
                }
            }
        }
    }

    private(set) var speakerVideoTrack: HMSVideoTrack? {
        didSet {
            if let oldValue = oldValue, let speakerVideoTrack = speakerVideoTrack {

                if let oldIndex = model.firstIndex(where: { $0.videoTrack.trackId == oldValue.trackId }),
                   let newIndex = model.firstIndex(where: { $0.videoTrack.trackId == speakerVideoTrack.trackId }) {
                    if oldIndex != newIndex {
                        model[oldIndex].isCurrentSpeaker = false
                        model[newIndex].isCurrentSpeaker = true
                        updateView(.refresh(indexes: (oldIndex, newIndex)))
                    }
                }
            }
        }
    }

    private var codec: HMSVideoCodec {
        let codecString = UserDefaults.standard.string(forKey: Constants.videoCodec) ?? "VP8"

        switch codecString {
        case "H264":
            return .H264
        case "VP9":
            return .VP9
        default:
            return .VP8
        }
    }

    private var resolution: HMSVideoResolution {
        let resolutionString = UserDefaults.standard.string(forKey: Constants.videoResolution) ?? "QHD"

        switch resolutionString {
        case "QVGA":
            return .QVGA
        case "VGA":
            return .VGA
        case "HD":
            return .HD
        case "Full HD":
            return .fullHD
        default:
            return .QHD
        }
    }

    private var cameraSource = "Front Facing" {
        willSet {
            if newValue != cameraSource {
                videoCapturer?.switchCamera()
            }
        }
    }

    internal var broadcasts = [[AnyHashable: Any]]()

    private var flow: MeetingFlow

    // MARK: - Setup Stream

    init(for user: String, in room: String, _ flow: MeetingFlow, _ callback: @escaping (VideoCellState) -> Void) {

        self.user = user
        self.flow = flow
        self.updateView = callback

        setup(room)

        observeSettingsUpdated()
    }

    func setup(_ room: String) {

        switch flow {
        case .join:
            fetchToken(Constants.endpoint, Constants.getTokenURL, room) { [weak self] token, error in

                guard error == nil, let token = token, let strongSelf = self
                else {
                    let error = error ?? CustomError(title: "Fetch Token Error")
                    NotificationCenter.default.post(name: Constants.hmsError,
                                                    object: nil,
                                                    userInfo: ["Error": error])
                    return
                }
                strongSelf.connect(with: token, room)
            }
        case .start:
            createRoom(name: room) { [weak self] _, roomID, error in

                guard error == nil, let roomID = roomID, let strongSelf = self
                else {
                    let error = error ?? CustomError(title: "Create Room Error")
                    NotificationCenter.default.post(name: Constants.hmsError,
                                                    object: nil,
                                                    userInfo: ["Error": error])
                    return
                }

                strongSelf.fetchToken(Constants.endpoint, Constants.getTokenURL, roomID) { [weak self] token, error in
                    guard error == nil, let token = token, let strongSelf = self
                    else {
                        let error = error ?? CustomError(title: "Fetch Token Error")
                        NotificationCenter.default.post(name: Constants.hmsError,
                                                        object: nil,
                                                        userInfo: ["Error": error])
                        return
                    }

                    strongSelf.connect(with: token, roomID)
                }
            }
        }
    }

    func fetchToken(_ endpoint: String,
                    _ token: String,
                    _ roomID: String,
                    completion: @escaping (String?, Error?) -> Void) {

        guard let endpointURL = URL(string: endpoint),
              let tokenURL = URL(string: token),
              let subDomain = endpointURL.host?.components(separatedBy: ".").first
        else {
            completion(nil, CustomError(title: Constants.urlEmpty))
            return
        }

        let body = [  "room_id": roomID,
                      "user_name": user,
                      "role": "guest",
                      "env": subDomain    ]

        var request = URLRequest(url: tokenURL)
        request.httpMethod = "POST"

        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body, options: .prettyPrinted)
        } catch {
            print(error.localizedDescription)
            completion(nil, CustomError(title: error.localizedDescription))
            return
        }

        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue("application/json", forHTTPHeaderField: "Accept")

        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in

            print(#function, response)
            self?.parseToken(from: data, error: error) { token, error in
                DispatchQueue.main.async {
                    completion(token, error)
                }
            }
        }.resume()
    }

    func parseToken(from data: Data?, error: Error?, completion: @escaping (String?, Error?) -> Void) {

        guard error == nil, let data = data else {
            let message = error?.localizedDescription ?? "Parse Token Error"
            completion(nil, CustomError(title: message))
            return
        }

        do {
            if let json = try JSONSerialization.jsonObject(with: data,
                                                           options: .mutableContainers) as? [String: Any] {
                if let token = json["token"] as? String {
                    completion(token, nil)
                } else {
                    print(Constants.jsonError)
                    completion(nil, CustomError(title: Constants.jsonError))
                }
            }
        } catch {
            print(error.localizedDescription)
            completion(nil, CustomError(title: error.localizedDescription))
        }
    }

    func createRoom(name: String, completion: @escaping (Bool, String?, Error?) -> Void) {

        guard let createRoomURL = URL(string: Constants.createRoomURL)
        else {
            completion(false, nil, CustomError(title: "Create Room URL Error"))
            return
        }

        let cleanedName = name.replacingOccurrences(of: " ", with: "")

        let body = [
                "room_id": cleanedName,
                "user_name": user,
                "role": "host",
                "env": "prod-in"
                ]

        var request = URLRequest(url: createRoomURL)
        request.httpMethod = "POST"

        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body, options: .prettyPrinted)
        } catch {
            print(error.localizedDescription)
            completion(false, nil, CustomError(title: error.localizedDescription))
            return
        }

        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue("application/json", forHTTPHeaderField: "Accept")

        URLSession.shared.dataTask(with: request) { [weak self] data, _, error in

            self?.parseRooms(data, error) { isSucces, roomID, error in
                completion(isSucces, roomID, error)
            }
        }.resume()
    }

    func parseRooms(_ data: Data?, _ error: Error?, completion: (Bool, String?, Error?) -> Void) {
        guard error == nil, let data = data else {
            let message = error?.localizedDescription ?? "Parse Create Room Response Error"
            completion(false, nil, CustomError(title: message))
            return
        }

        do {
            if let json = try JSONSerialization.jsonObject(with: data,
                                                           options: .mutableContainers) as? [String: Any] {
                if let roomID = json["id"] as? String {
                    completion(true, roomID, nil)
                } else {
                    print(Constants.jsonError)
                    completion(false, nil, CustomError(title: Constants.jsonError))
                }
            }
        } catch {
            print(error.localizedDescription)
            completion(false, nil, CustomError(title: error.localizedDescription))
        }
    }

    // MARK: - Stream Handlers

    func connect(with token: String, _ roomID: String) {

        localPeer = HMSPeer(name: user, authToken: token)

        let config = HMSClientConfig()
        config.endpoint = Constants.endpoint

        client = HMSClient(peer: localPeer, config: config)
        client.logLevel = .verbose

        room = HMSRoom(roomId: roomID)

        setupCallbacks()

        setAudioDelay()

        client.connect()
    }

    func setupCallbacks() {
        client.onPeerJoin = { room, peer in
            print("onPeerJoin: ", room.roomId, peer.name)
            NotificationCenter.default.post(name: Constants.peersUpdated, object: nil)
        }

        client.onPeerLeave = { room, peer in
            print("onPeerLeave: ", room.roomId, peer.name)
            NotificationCenter.default.post(name: Constants.peersUpdated, object: nil)
        }

        client.onStreamAdd = { [weak self] room, peer, info in
            print("onStreamAdd: ", room.roomId, peer.name, info.streamId)
            self?.subscribe(to: room, peer, with: info)
        }

        client.onStreamRemove = { [weak self] room, peer, info in

            print("onStreamRemove: ", room.roomId, peer.name, info.streamId)

            if let videoTracks = self?.videoTracks {

                for (index, track) in videoTracks.enumerated() where track.streamId == info.streamId {
                    self?.videoTracks.remove(at: index)
                }

                var indexes = [Int]()
                if let model = self?.model {
                    for (index, item) in model.enumerated() where item.videoTrack.streamId == info.streamId {
                        self?.model.remove(at: index)
                        indexes.append(index)
                    }
                }

                indexes.forEach { self?.updateView(.delete(index: $0)) }
            }
        }

        client.onBroadcast = { [weak self] room, peer, data in
            print("onBroadcast: ", room.roomId, peer.peerId, data)
            self?.broadcasts.append(data)
            NotificationCenter.default.post(name: Constants.broadcastReceived, object: nil)
        }

        client.onConnect = { [weak self] in
            self?.client.join((self?.room)!) { isSuccess, error in
                print("client.join: ", isSuccess, error ?? "No Error")
                self?.publish()
            }
        }

        client.onDisconnect = { error in

            let message = error?.localizedDescription ?? "Client disconnected!"

            NotificationCenter.default.post(name: Constants.hmsError,
                                            object: nil,
                                            userInfo: ["error": message])
        }

        client.onAudioLevelInfo = { [weak self] levels in
            self?.updateAudio(with: levels)
        }
    }

    func subscribe(to room: HMSRoom, _ peer: HMSPeer, with info: HMSStreamInfo) {

        peers[info.streamId] = peer

        client.subscribe(info, room: room) { [weak self] (stream, error) in

            guard let stream = stream,
                  let videoTrack = stream.videoTracks?.first
            else {
                print(error?.localizedDescription ?? "Client Subscribe Error")
                return
            }

            self?.remoteStreams.append(stream)
            self?.videoTracks.append(videoTrack)

            let item = VideoModel(peer, stream, videoTrack)
            self?.model.append(item)
            self?.updateView(.insert(index: (self?.model.count ?? 1) - 1))
        }
    }

    func publish() {

        let userDefaults = UserDefaults.standard

        let constraints = HMSMediaStreamConstraints()
        constraints.shouldPublishAudio = userDefaults.object(forKey: Constants.publishAudio) as? Bool ?? true
        constraints.shouldPublishVideo = userDefaults.object(forKey: Constants.publishVideo) as? Bool ?? true
        constraints.bitrate = userDefaults.object(forKey: Constants.videoBitRate) as? Int ?? 256
        constraints.audioBitrate = userDefaults.object(forKey: Constants.audioBitRate) as? Int ?? 0
        constraints.frameRate = userDefaults.object(forKey: Constants.videoFrameRate) as? Int ?? 25
        constraints.resolution = resolution
        constraints.codec = codec

        guard let localStream = try? client.getLocalStream(constraints) else {
            return
        }

        peers[localStream.streamId] = localPeer

        let audioPollDelay = userDefaults.object(forKey: Constants.audioPollDelay) as? Double ?? 0.5
        client.startAudioLevelMonitor(audioPollDelay)

        client.publish(localStream, room: room) { stream, error in
            guard let stream = stream else {
                print(error?.localizedDescription ?? "Local Stream publish failed")
                return
            }

            self.setupLocal(stream)
        }
    }

    func setupLocal(_ stream: HMSStream) {
        localStream = stream
        videoCapturer = stream.videoCapturer
        localAudioTrack = stream.audioTracks?.first
        localVideoTrack = stream.videoTracks?.first

        if let source = UserDefaults.standard.string(forKey: Constants.defaultVideoSource) {
            if source == "Rear Facing" {
                videoCapturer?.switchCamera()
                cameraSource = "Rear Facing"
            } else {
                cameraSource = "Front Facing"
            }
        }

        videoCapturer?.startCapture()

        if let track = localVideoTrack {
            videoTracks.append(track)

            guard let peer = peers[stream.streamId] else {
                print("Error: Could not find local peer!")
                return
            }

            let item = VideoModel(peer, stream, track)
            model.append(item)

            let lastIndex = model.count > 0 ? model.count : 1
            updateView(.insert(index: lastIndex - 1))
        }
    }
}

// MARK: - Action Handlers

extension HMSInteractor {

    func updateAudio(with levels: [HMSAudioLevelInfo]) {

        guard let topLevel = levels.first,
              let peer = peers[topLevel.streamId]
        else {
            return
        }

        speakerPeerID = peer.peerId

        print("Speaker: ", peer.name)
    }

    func observeSettingsUpdated() {
        _ = NotificationCenter.default.addObserver(forName: Constants.settingsUpdated,
                                                   object: nil,
                                                   queue: .main) { [weak self] _ in

            let userDefaults = UserDefaults.standard

            let constraints = HMSMediaStreamConstraints()

            constraints.bitrate = userDefaults.object(forKey: Constants.videoBitRate) as? Int ?? 256
            constraints.audioBitrate = userDefaults.object(forKey: Constants.audioBitRate) as? Int ?? 0
            constraints.frameRate = userDefaults.object(forKey: Constants.videoFrameRate) as? Int ?? 25
            constraints.resolution = self?.resolution ?? .QHD

            do {
                if let strongSelf = self, let stream = self?.localStream {
                    try strongSelf.client.applyConstraints(constraints, to: stream)
                }
            } catch {
                NotificationCenter.default.post(name: Constants.hmsError, object: nil, userInfo: ["Error": error])
            }

            if let source = UserDefaults.standard.string(forKey: Constants.defaultVideoSource) {
                self?.cameraSource = source
            }
            
            let publishVideo = UserDefaults.standard.object(forKey: Constants.publishVideo) as? Bool ?? true
            self?.localStream?.videoTracks?.first?.enabled = publishVideo
            
            let publishAudio = UserDefaults.standard.object(forKey: Constants.publishAudio) as? Bool ?? true
            self?.localStream?.audioTracks?.first?.enabled = publishAudio

            self?.setAudioDelay()
        }
    }

    func setAudioDelay() {
        let audioPollDelay = UserDefaults.standard.object(forKey: Constants.audioPollDelay) as? Double ?? 3.0
        client.startAudioLevelMonitor(audioPollDelay)
    }

    func send(_ broadcast: [AnyHashable: Any], completion: @escaping () -> Void) {

        client.broadcast(broadcast, room: room) { _, error in

            if let error = error {
                print(error.localizedDescription)
            }
            completion()
        }
    }

    func cleanup() {
        guard let client = client else {
            return
        }

        videoCapturer?.stopCapture()

        client.leave(room)

        client.disconnect()
    }
}
